eth/api_tracer.go

```diff
@@ -60,9 +60,10 @@ const (
 // TraceConfig holds extra parameters to trace functions.
 type TraceConfig struct {
 	*vm.LogConfig
-	Tracer  *string
-	Timeout *string
-	Reexec  *uint64
+	Tracer                 *string
+	Timeout                *string
+	Reexec                 *uint64
+	ParityCompatibleOutput bool
```

Comment:
Does `TraceConfig` have custom JSON Un/Marshal methods? Where/how are these fields configured? (Forgive my naivete... :blush: )

Reply:
No, it doesn't. It just uses the default field names as the name. They just fallback to the default value for the type at the moment, or if no key is set, then the whole config is nil.

I was thinking to make this explicit when I joined ETC and started working on Tracers, but left it like this to not trigger any future conflicts with go-ethereum.

Reply:
Related, here I initiate it using the default if needed. https://github.com/etclabscore/core-geth/blob/feat/trace-state-diff/eth/api_tracer_parity.go#L66

Reply:
Got it. For ease of reference

```go
// setConfigTracerToParity forces the Tracer to the Parity one
func setConfigTracerToParity(config *TraceConfig) *TraceConfig {
	if config == nil {
		config = &TraceConfig{}
	}

	if config.Tracer == nil {
		tracer := "callTracerParity"
		config.Tracer = &tracer
	}

	return config
}
```

Used as 

```go
// Block returns the structured logs created during the execution of
// EVM and returns them as a JSON object.
// The correct name will be TraceBlockByNumber, though we want to be compatible with Parity trace module.
func (api *PrivateTraceAPI) Block(ctx context.Context, number rpc.BlockNumber, config *TraceConfig) ([]interface{}, error) {
	// Fetch the block that we want to trace
	var block *types.Block

	switch number {
	case rpc.PendingBlockNumber:
		block = api.eth.miner.PendingBlock()
	case rpc.LatestBlockNumber:
		block = api.eth.blockchain.CurrentBlock()
	default:
		block = api.eth.blockchain.GetBlockByNumber(uint64(number))
	}
	// Trace the block if it was found
	if block == nil {
		return nil, fmt.Errorf("block #%d not found", number)
	}

	config = setConfigTracerToParity(config)

	traceResults, err := traceBlockByNumber(ctx, api.eth, number, config)
	if err != nil {
		return nil, err
	}
```

In this case it seems to me like the `setConfigTracerToParity` is actually _defaulting_ the config to Parity in case the value is nil (as opposed to `forces the Tracer to the Parity one`), or? Maybe just a rename for that function?

Reply:
I'm curious about the JSON field encoding here. Normally JSON fields for exported methods will use capitalized field names;

```go
//cat main.go
package main

import (
        "encoding/json"
        "fmt"
)

type MyStruct struct {
        Reexec             *string
        ParityCompatOutput bool
}

func main() {
        ex := &MyStruct{}

        b, _ := json.MarshalIndent(ex, "", "    ")
        fmt.Println(string(b))
}
```

```sh
go run main.go
{
    "Reexec": null,
    "ParityCompatOutput": false
}
```

And in the internal/web3ext.go it doesn't look like this arg is getting a custom formatter. (Also, should `callMany` have same formatters as `call`?).  

```js
		new web3._extend.Method({
			name: 'call',
			call: 'trace_call',
			params: 3,
			inputFormatter: [web3._extend.formatters.inputCallFormatter, null, null]
		}),
		new web3._extend.Method({
			name: 'callMany',
			call: 'trace_callMany',
			params: 3,
			inputFormatter: [null, null, null]
		}),
```

Reply:
> In this case it seems to me like the setConfigTracerToParity is actually defaulting the config to Parity in case the value is nil (as opposed to forces the Tracer to the Parity one), or? Maybe just a rename for that function?

Good catch. It's a comment issue. It has been left off, from the initial tracing work. Will be updated.

> I'm curious about the JSON field encoding here. Normally JSON fields for exported methods will use capitalised field names;

True. Currently this is only used to Unmarshal JSON and the default parser handles `camelCase` as well. I followed go-ethereum practice without paying much attention to this one and you are right that this can create issues later on.
Do you think it's good to handle this now? and if yes, shall we do this in this or another PR?

> And in the internal/web3ext.go it doesn't look like this arg is getting a custom formatter. (Also, should callMany have same formatters as call? No, because args are different; callMany uses an array of CallArgs. Should this have its own formatter?).

Yes, this is a great idea. Though we have to check out how the `internal/jsre/deps/web3.js` is being packaged externally, and if such change will be easy to maintain (conflicts resolution) later on.
@meowsbits idea to see if we can extend this functionality in `web3ext.go` is great. I will examine this, as I have some concerns with regards the custom function context.

Reply:
Resolved by e0b8ae0 and 84bf302.

Reply:
Nicely done with https://github.com/etclabscore/core-geth/pull/345/commits/e0b8ae0ce39231bf76229dbbf5142aa496bdd139!

Reply:
> I followed go-ethereum practice without paying much attention to this one and you are right that this can create issues later on. Do you think it's good to handle this now? and if yes, shall we do this in this or another PR?

Understood. As long as there are tests ensuring that all the fields of the Config get decoded as expected, then I think it's fine as-is.